### ISSUE TITLE
Kubernetes ingress missing url

### ISSUE THREAD

**Issue by NissesSenap** [grafana-operator/grafana-operator-experimental#40]
I'm having issues trying to create grafana instance in kubernetes when using `make run`

I have created a simple grafana instance using the example files with some extra additions.

```.yaml
---
apiVersion: grafana.integreatly.org/v1beta1
kind: Grafana
metadata:
  name: grafana
  labels:
    dashboards: "grafana"
spec:
  client:
    preferIngress: true
  config:
    log:
      mode: "console"
    auth:
      disable_login_form: "false"
    security:
      admin_user: root
      admin_password: secret
  ingress:
    spec:
      ingressClassName: nginx
      rules:
        - host: test.io
          http:
            paths:
              - backend:
                  service:
                    name: grafana-service
                    port:
                      number: 3000
                path: /
                pathType: Prefix
---
apiVersion: grafana.integreatly.org/v1beta1
kind: GrafanaDashboard
metadata:
  name: grafana-dashboard
spec:
  instanceSelector:
    matchLabels:
      dashboards: "grafana"
  json: >
    {
      "id": null,
      "title": "Simple Dashboard",
      "tags": [],
      "style": "dark",
      "timezone": "browser",
      "editable": true,
      "hideControls": false,
      "graphTooltip": 1,
      "panels": [],
      "time": {
        "from": "now-6h",
        "to": "now"
      },
      "timepicker": {
        "time_options": [],
        "refresh_intervals": []
      },
      "templating": {
        "list": []
      },
      "annotations": {
        "list": []
      },
      "refresh": "5s",
      "schemaVersion": 17,
      "version": 0,
      "links": []
    }
---
apiVersion: grafana.integreatly.org/v1beta1
kind: GrafanaDatasource
metadata:
  name: example-grafanadatasource
spec:
  datasource:
    access: proxy
    database: prometheus
    jsonData:
      timeInterval: 5s
      tlsSkipVerify: true
    name: Prometheus
    url: http://prometheus-service:9090
  instanceSelector:
    matchLabels:
      dashboards: "grafana"
```

```.txt
1.6686266712433786e+09  INFO    found matching Grafana instances for datasource {"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"example-grafanadatasource","namespace":"grafana-operator-system"}, "namespace": "grafana-operator-system", "name": "example-grafanadatasource", "reconcileID": "44852e11-9b41-4be7-9f23-3594cc4a6544", "count": 1}
1.6686266712433953e+09  INFO    grafana instance not ready      {"controller": "grafanadatasource", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDatasource", "GrafanaDatasource": {"name":"example-grafanadatasource","namespace":"grafana-operator-system"}, "namespace": "grafana-operator-system", "name": "example-grafanadatasource", "reconcileID": "44852e11-9b41-4be7-9f23-3594cc4a6544", "grafana": "grafana"}
1.6686267822545404e+09  INFO    found matching Grafana instances for dashboard  {"controller": "grafanadashboard", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDashboard", "GrafanaDashboard": {"name":"grafana-dashboard","namespace":"grafana-operator-system"}, "namespace": "grafana-operator-system", "name": "grafana-dashboard", "reconcileID": "44d02bbc-3696-4d19-9839-2f2b67f0b609", "count": 1}
1.668626782254582e+09   INFO    grafana instance not ready      {"controller": "grafanadashboard", "controllerGroup": "grafana.integreatly.org", "controllerKind": "GrafanaDashboard", "GrafanaDashboard": {"name":"grafana-dashboard","namespace":"grafana-operator-system"}, "namespace": "grafana-operator-system", "name": "grafana-dashboard", "reconcileID": "44d02bbc-3696-4d19-9839-2f2b67f0b609", "grafana": "grafana"}
```

I assume this is due to the missing of URL in the status:

```
status:
  stage: complete
  stageStatus: success
```

But if I instead use the example grafana straight up and don't define any custom ingress config:

```.yaml
---
apiVersion: grafana.integreatly.org/v1beta1
kind: Grafana
metadata:
  name: grafana
  labels:
    dashboards: "grafana"
spec:
  client:
    preferIngress: true
  config:
    log:
      mode: "console"
    auth:
      disable_login_form: "false"
    security:
      admin_user: root
      admin_password: secret
``` 

Then I get

``` 
status:
  adminUrl: https://
  stage: complete
  stageStatus: success
```

It might be something dumb how I have setup my local env.
Any pointers would be great.

**Comment by NissesSenap** [comment]
I solved it with an ugly workaround using the ip of my kind cluster
`docker container inspect kind-control-plane   --format '{{ .NetworkSettings.Networks.kind.IPAddress }}'`

This will provide you with an IP.
I then put this IP in to my /etc/hosts file
```/etc/hosts
172.18.0.2 grafana-service.grafana-operator-system.svc.cluster.local
```

And then finally force my grafana service to use this ip.

``` 
apiVersion: grafana.integreatly.org/v1beta1
kind: Grafana
metadata:
  name: grafana
  labels:
    dashboards: "grafana"
spec:
  client:
    preferIngress: false
  config:
    log:
      mode: "console"
    auth:
      disable_login_form: "false"
    security:
      admin_user: root
      admin_password: secret
  service:
    spec:
      externalIPs:
        - 172.18.0.2
``` 

But this workaround shouldn't be needed. I have a ingress controller on my kind cluster and I think the grafana controller should be able to find my hostname that I tried to set in the ingress config file.
I will try to read some more code when I have time.

**Comment by pb82** [comment]
@NissesSenap Ok, I need to try this with kind. When using k3d, I end up with the Ingress IP in the `status.adminUrl` which works.

We're currently reading the `host` from `status.loadbalancer`. I'm not sure if this is the best or even  correct way to do it: https://github.com/grafana-operator/grafana-operator-experimental/blob/master/controllers/reconcilers/grafana/ingress_reconciler.go#L68

Can you attach the Ingress yaml as well? Would be good to see what your Ingress looks like and what we can work with.

**Comment by weisdd** [comment]
@NissesSenap normally, in non-OpenShift environments it's enough to pass "preferIngress: false".
As for "preferIngress: true", you can either use services like nip.io (e.g. grafana.192.168.64.2.nip.io, so it gets resolved into 192.168.64.2 through any DNS) or, if using aliases, you need to pass hostAliases.

**Comment by NissesSenap** [comment]
Right I know I have looked at this earlier.
So the funny thing is that the value looks okay according to your logic, this is the ingress object that get's generated.

``` apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: grafana-ingress
  namespace: grafana-operator-system
  ownerReferences:
  - apiVersion: grafana.integreatly.org/v1beta1
    kind: Grafana
    name: grafana
    uid: 0ae0bba4-7383-4534-be98-8582b20a8272
  resourceVersion: "2906135"
  uid: 2952ec2d-892e-49de-b7f4-34113aa5ef0b
spec:
  rules:
  - http:
      paths:
      - backend:
          service:
            name: grafana-service
            port:
              number: 3000
        path: /
        pathType: Prefix
status:
  loadBalancer:
    ingress:
    - hostname: localhost
```

**Comment by NissesSenap** [comment]
I have tried a few things back and forth.
It seems to me like the initial time I'm running the make command and my resources don't exist they don't go in to the if statment `if cr.PreferIngress() {`.

But if I restart my operator that I run locally it jumps in to it.
But for some reason it don't want to use the string from `ingress.Hostname` and I can't wrap my head around it right now....

@pb82 if you have time some day we could do some pair debugging on this on a kind cluster.
I also realized an issue, we can't assume that the ingress is https://, there is nothing that stops the user from creating a ingress and want to prefer that to be run on http. It's not a good idea but it can be done. 
I will create a separate issue for that.
